src/fparser/two/tests/utils/test_blockbase.py

```diff
@@ -87,3 +88,96 @@ def test_include(f2003_create):
         "'! comment5')), End_Program_Stmt('PROGRAM', Name('test'))],)") \
         in str(result)
     assert "should" not in str(result)
+
+
+@pytest.mark.parametrize('strict_order', [True, False])
+def test_strict_order_invalid_code(f2003_create, strict_order):
+    '''Check that the strict_order flag toggles the parse behaviour as
+    expected.
+    '''
+    subclasses = [F2003.Specification_Part, F2003.Execution_Part]
+    reader = get_reader("""
+        program main
+            i = 2
+            integer :: i
+        end program main
+        """)
+
+    expected = remove_indentation("""([
+      Program_Stmt('PROGRAM', Name('main')),
+      Execution_Part(
+          Assignment_Stmt(Name('i'), '=',
+          Int_Literal_Constant('2', None))),
+      Specification_Part(
+          Type_Declaration_Stmt(
+              Intrinsic_Type_Spec('INTEGER', None),
+              None,
+              Entity_Decl(Name('i'), None, None, None))),
+      End_Program_Stmt('PROGRAM', Name('main'))
+     ],)
+    """)
+    result = BlockBase.match(
+        F2003.Program_Stmt, subclasses, F2003.End_Program_Stmt, reader,
+        strict_order=strict_order)
+
+    if strict_order:
+        assert result == None
+    else:
+        assert str(result) == expected
+
+
+def test_strict_order_valid_code(f2003_create):
+    '''Tests that the strict_order keyword allows repeated types.
+    '''
+    subclasses = [F2003.Specification_Part, F2003.Execution_Part]
+    reader = get_reader("""
+        program main
+            integer :: i
+            real    :: rho
+            i = 2
+            rho = i * 3.14
+        end program main
+        """)
+
+    expected = remove_indentation("""([
+      Program_Stmt('PROGRAM', Name('main')),
+      Specification_Part(
+          Type_Declaration_Stmt(
+              Intrinsic_Type_Spec('INTEGER', None), None,
+              Entity_Decl(Name('i'), None, None, None)),
+          Type_Declaration_Stmt(
+              Intrinsic_Type_Spec('REAL', None), None,
+              Entity_Decl(Name('rho'), None, None, None))),
+      Execution_Part(
+          Assignment_Stmt(Name('i'), '=', Int_Literal_Constant('2', None)),
+          Assignment_Stmt(
+              Name('rho'), '=', Add_Operand(Name('i'), '*',
+              Real_Literal_Constant('3.14', None)))),
+      End_Program_Stmt('PROGRAM', Name('main'))
+     ],)
+    """)
+    result = BlockBase.match(
+        F2003.Program_Stmt, subclasses, F2003.End_Program_Stmt, reader,
+        strict_order=True)
+
+    assert str(result) == expected
+
+
+def remove_indentation(string):
+    '''
+    A utility function that removes indented multiline strings of reprs.
+
```

Comment:
utilities (even ones in tests) need their interfaces specified in docstrings I'm afraid.

Reply:
The interface is the wrong form, but I'll change it on master with the changelog as this PR has iterated enough.